### PR TITLE
Fix issue https://github.com/Microsoft/TypeScript/issues/3277

### LINKED ISSUE #3277: `tsc --watch` recreates deleted source files
In current stable and also master (de88ede4feeabab91d032359406a19166ed5bbb2), `tsc --watch` will keep recreating `file.js` during recompilation even when the original `file.ts` source file is removed. Only workaround is to restart `tsc --watch` to make it forget about `file.ts`.

For example,

```
$ ls incorrect_recreation/
test2.ts  test.ts  tsconfig.json
$ cat tsconfig.json
<empty>
```
```
$ node bin/tsc.js --watch -p incorrect_recreation/
message TS6042: Compilation complete. Watching for file changes.
```
In another terminal:
```
$ ls
test2.js  test2.ts  test.js  test.ts  tsconfig.json # as expected
$ rm test.ts test.js 
$ ls
test2.js  test2.ts  tsconfig.json # so, no trace of test.js or test.ts now
$ touch test2.ts # force a recompile and wait
$ ls
test2.js  test2.ts  test.js  tsconfig.json # test.js should not have been recreated
```

### PR DESCRIPTION
**Pull request by ShyykoSerhiy**
Fixes #3277.

### CODE REVIEW

File: src/compiler/sys.ts [hunk 1]
```diff
@@ -282,11 +282,16 @@ namespace ts {
                     };
 
                     function fileChanged(curr: any, prev: any) {
+                        // mtime.getTime() equals 0 if file was removed
+                        if (curr.mtime.getTime() === 0) {
+                            callback(fileName, true);
```
Comment (RyanCavanaugh) [inline src/compiler/sys.ts]: Add comment with parameter name in front of boolean arguments (e.g. `callback(fileName, /* removed */ true);`

File: src/compiler/tsc.ts [hunk 2]
```diff
@@ -279,9 +279,15 @@ namespace ts {
         }
 
         // If a source file changes, mark it as unwatched and start the recompilation timer
-        function sourceFileChanged(sourceFile: SourceFile) {
+        function sourceFileChanged(sourceFile: SourceFile, removed: boolean) {
             sourceFile.fileWatcher.close();
             sourceFile.fileWatcher = undefined;
+            if (removed){
```
Comment (DanielRosenwasser) [inline src/compiler/tsc.ts]: Space between `){`

File: src/compiler/tsc.ts [hunk 3]
```diff
@@ -279,9 +279,15 @@ namespace ts {
         }
 
         // If a source file changes, mark it as unwatched and start the recompilation timer
-        function sourceFileChanged(sourceFile: SourceFile) {
+        function sourceFileChanged(sourceFile: SourceFile, removed: boolean) {
             sourceFile.fileWatcher.close();
             sourceFile.fileWatcher = undefined;
+            if (removed){
+                var index = rootFileNames.indexOf(sourceFile.fileName);
+                if (index !== -1){
```
Comment (DanielRosenwasser) [inline src/compiler/tsc.ts]: Space between `){`

Comment (DanielRosenwasser) [inline src/compiler/tsc.ts]: `index >= 0`

File: src/compiler/tsc.ts [hunk 4]
```diff
@@ -279,9 +279,15 @@ namespace ts {
...
 
         // If a source file changes, mark it as unwatched and start the recompilation timer
-        function sourceFileChanged(sourceFile: SourceFile) {
+        function sourceFileChanged(sourceFile: SourceFile, removed: boolean) {
             sourceFile.fileWatcher.close();
             sourceFile.fileWatcher = undefined;
+            if (removed){
+                var index = rootFileNames.indexOf(sourceFile.fileName);
+                if (index !== -1){
+                    rootFileNames.splice(index, 1);
```
Comment (DanielRosenwasser) [inline src/compiler/tsc.ts]: Does this cover referenced files?

Also, I don't know if this is the right approach. I think that if a root file is deleted, we should be removing the file from the cache, but we should be erroring when we fail to find it.

**Comment by mhegazy** [comment]
@ShyykoSerhiy can you refresh and resubmit?

**Comment by mhegazy** [comment]
:+1:

**Comment by mhegazy** [comment]
Sorry for the delay. merged manually.

**Pull request merged**